Fails to build on Python 3.10.0a2

Trying to build zodbpickle on Python 3.10.0a2 fails with

```
  gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/home/mg/src/zodbbrowser/.tox/py310/include -I/home/mg/opt/python310/include/python3.10 -c src/zodbpickle/_pickle_33.c -o build/temp.linux-x86_64-3.10/src/zodbpickle/_pickle_33.o
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_New’:
  src/zodbpickle/_pickle_33.c:179:19: error: lvalue required as left operand of assignment
    179 |     Py_SIZE(self) = 0;
        |                   ^
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_clear’:
  src/zodbpickle/_pickle_33.c:205:19: error: lvalue required as left operand of assignment
    205 |     Py_SIZE(self) = clearto;
        |                   ^
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_pop’:
  src/zodbpickle/_pickle_33.c:247:23: error: lvalue required as decrement operand
    247 |     return self->data[--Py_SIZE(self)];
        |                       ^~
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_push’:
  src/zodbpickle/_pickle_33.c:257:29: error: lvalue required as increment operand
    257 |     self->data[Py_SIZE(self)++] = obj;
        |                             ^~
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_poptuple’:
  src/zodbpickle/_pickle_33.c:283:19: error: lvalue required as left operand of assignment
    283 |     Py_SIZE(self) = start;
        |                   ^
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_poplist’:
  src/zodbpickle/_pickle_33.c:300:19: error: lvalue required as left operand of assignment
    300 |     Py_SIZE(self) = start;
        |                   ^
  src/zodbpickle/_pickle_33.c: In function ‘save_long’:
  src/zodbpickle/_pickle_33.c:1646:18: warning: implicit declaration of function ‘_PyUnicode_AsStringAndSize’; did you mean ‘PyUnicode_FromStringAndSize’? [-Wimplicit-function-declaration]
   1646 |         string = _PyUnicode_AsStringAndSize(repr, &size);
        |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
        |                  PyUnicode_FromStringAndSize
  src/zodbpickle/_pickle_33.c:1646:16: warning: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   1646 |         string = _PyUnicode_AsStringAndSize(repr, &size);
        |                ^
  src/zodbpickle/_pickle_33.c: In function ‘save_pers’:
  src/zodbpickle/_pickle_33.c:2942:29: warning: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   2942 |             pid_ascii_bytes = _PyUnicode_AsStringAndSize(pid_str, &size);
        |                             ^
  src/zodbpickle/_pickle_33.c: In function ‘Pickler_init’:
  src/zodbpickle/_pickle_33.c:3582:9: warning: implicit declaration of function ‘_PyObject_HasAttrId’; did you mean ‘_PyObject_SetAttrId’? [-Wimplicit-function-declaration]
   3582 |     if (_PyObject_HasAttrId((PyObject *)self, &PyId_persistent_id)) {
        |         ^~~~~~~~~~~~~~~~~~~
        |         _PyObject_SetAttrId
  src/zodbpickle/_pickle_33.c: In function ‘load_pop’:
  src/zodbpickle/_pickle_33.c:4840:30: error: lvalue required as left operand of assignment
   4840 |         Py_SIZE(self->stack) = len;
        |                              ^
  src/zodbpickle/_pickle_33.c: In function ‘do_append’:
  src/zodbpickle/_pickle_33.c:5145:38: error: lvalue required as left operand of assignment
   5145 |                 Py_SIZE(self->stack) = x;
        |                                      ^
  src/zodbpickle/_pickle_33.c:5151:30: error: lvalue required as left operand of assignment
   5151 |         Py_SIZE(self->stack) = x;
        |                              ^
  src/zodbpickle/_pickle_33.c: In function ‘Pdata_pop’:
  src/zodbpickle/_pickle_33.c:248:1: warning: control reaches end of non-void function [-Wreturn-type]
    248 | }
        | ^
  error: command '/usr/lib/ccache/gcc' failed with exit code 1
```

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -103,6 +103,7 @@ jobs:
           - 3.7
           - 3.8
           - 3.9
+          - 3.10.0-beta.3
         os: [ubuntu-20.04, macos-latest]
         exclude:
           - os: macos-latest
@@ -189,6 +190,7 @@ jobs:
           - 3.7
           - 3.8
           - 3.9
+          - 3.10.0-beta.3
         os: [ubuntu-20.04, macos-latest]
         exclude:
           - os: macos-latest

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -103,7 +103,7 @@ jobs:
           - 3.7
           - 3.8
           - 3.9
-          - 3.10.0-beta.3
+          - 3.10.0-rc.2
         os: [ubuntu-20.04, macos-latest]
         exclude:
           - os: macos-latest
@@ -190,7 +190,7 @@ jobs:
           - 3.7
           - 3.8
           - 3.9
-          - 3.10.0-beta.3
+          - 3.10.0-rc.2
         os: [ubuntu-20.04, macos-latest]
         exclude:
           - os: macos-latest

--- a/.manylinux-install.sh
+++ b/.manylinux-install.sh
@@ -33,6 +33,7 @@ for PYBIN in /opt/python/*/bin; do
        [[ "${PYBIN}" == *"cp36"* ]] || \
        [[ "${PYBIN}" == *"cp37"* ]] || \
        [[ "${PYBIN}" == *"cp38"* ]] || \
+       [[ "${PYBIN}" == *"cp310"* ]] || \
        [[ "${PYBIN}" == *"cp39"* ]]; then
         "${PYBIN}/pip" install -e /io/
         "${PYBIN}/pip" wheel /io/ -w wheelhouse/

--- a/.meta.toml
+++ b/.meta.toml
@@ -10,7 +10,7 @@ with-windows = false
 with-pypy = true
 with-legacy-python = true
 with-sphinx-doctests = false
-with-future-python = false
+with-future-python = true
 
 [tox]
 use-flake8 = false

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -5,7 +5,7 @@
 2.1.1 (unreleased)
 ==================
 
-- Nothing changed yet.
+- Add support for Python 3.10.
 
 
 2.1.0 (2021-09-24)

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -21,6 +21,10 @@ environment:
     - python: 38-x64
     - python: 39
     - python: 39-x64
+    # `multibuild` cannot install non-final versions as they are not on
+    # ftp.python.org, so we skip Python 3.10 until its final release:
+    # - python: 310
+    # - python: 310-x64
 
 install:
   - "SET PYTHONVERSION=%PYTHON%"

--- a/setup.py
+++ b/setup.py
@@ -67,6 +67,7 @@ setup(
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
         'Programming Language :: Python :: 3.9',
+        'Programming Language :: Python :: 3.10',
         'Programming Language :: Python :: Implementation :: CPython',
         'Programming Language :: Python :: Implementation :: PyPy',
         'Programming Language :: Python :: Implementation :: Jython',

--- a/src/zodbpickle/_pickle_33.c
+++ b/src/zodbpickle/_pickle_33.c
@@ -4,6 +4,28 @@
 PyDoc_STRVAR(pickle_module_doc,
 "Optimized C implementation for the Python pickle module.");
 
+
+#if (PY_VERSION_HEX >= 0x30A00B1) /* 3.10.0b1 */
+#ifndef Py_SIZE
+#define Py_SIZE(ob) (((PyVarObject*)(ob))->ob_size)
+#endif
+#define _PyUnicode_AsStringAndSize PyUnicode_AsUTF8AndSize
+/**
+ * The function ``_PyObject_LookupAttrId`` function replaces the combo of
+ * ``_PyObject_HasAttrId`` followed by ``_PyObject_GetAttrId``; our code isn't
+ * structured to take advantage of that, so for now we throw away the
+ * resulting attribute value and continue to make the extra ``Get`` call.
+ */
+static int _PyObject_HasAttrId(PyObject* obj, void* id)
+{
+    int result;
+    PyObject* attr_val;
+    result = _PyObject_LookupAttrId(obj, id, &attr_val);
+    Py_XDECREF(attr_val);
+    return result;
+}
+#endif
+
 /* Bump this when new opcodes are added to the pickle protocol. */
 enum {
     HIGHEST_PROTOCOL = 3,
@@ -1633,7 +1655,7 @@ save_long(PicklerObject *self, PyObject *obj)
             goto error;
     }
     else {
-        char *string;
+        const char *string;
 
         /* proto < 2: write the repr and newline.  This is quadratic-time (in
            the number of digits), in both directions.  We add a trailing 'L'
@@ -2929,7 +2951,7 @@ save_pers(PicklerObject *self, PyObject *obj, PyObject *func)
         }
         else {
             PyObject *pid_str = NULL;
-            char *pid_ascii_bytes;
+            const char *pid_ascii_bytes;
             Py_ssize_t size;
 
             pid_str = PyObject_Str(pid);

--- a/src/zodbpickle/tests/pickletester_3.py
+++ b/src/zodbpickle/tests/pickletester_3.py
@@ -1,14 +1,16 @@
 import io
+import os
 import unittest
 import sys
 import copyreg
 import weakref
+import tempfile
 from http.cookies import SimpleCookie
 from zodbpickle import pickle_3 as pickle
 from zodbpickle import pickletools_3 as pickletools
 
 from test.support import (
-    TestFailed, TESTFN, run_with_locale,
+    TestFailed, run_with_locale,
     _2G, _4G, bigmemtest,
     )
 
@@ -41,6 +43,8 @@ protocols = range(pickle.HIGHEST_PROTOCOL + 1)
 
 ascii_char_size = 1
 
+fd, TESTFN = tempfile.mkstemp('.pickletester_3')
+os.close(fd)
 
 # Return True if opcode code appears in the pickle, else False.
 def opcode_in_pickle(code, pickle):

--- a/tox.ini
+++ b/tox.ini
@@ -10,6 +10,7 @@ envlist =
     py37,py37-pure
     py38,py38-pure
     py39,py39-pure
+    py310,py310-pure
     pypy
     pypy3
     coverage

